include/dali/core/mm/monotonic_resource.h

```diff
@@ -0,0 +1,203 @@
+#include "dali/core/mm/detail/align.h"
+#include "dali/core/small_vector.h"
+
+namespace dali {
+namespace mm {
+
+/**
+ * @brief Monotonic resource which returns memory from a preallocated buffer.
+ *
+ * Monotonic buffer resources don't require manual deallocation of the returned pointers.
+ * The memory is freed in bulk when the underlying buffer is freed.
+ */
+class monotonic_buffer_resource : public memory_resource {
+ public:
+  monotonic_buffer_resource() = default;
+  monotonic_buffer_resource(void *memory, size_t bytes)
+  : curr_(static_cast<char*>(memory)), limit_(static_cast<char*>(memory) + bytes) {}
+
+  size_t avail(size_t alignment) const {
+    return limit_ - detail::align_ptr(curr_, alignment);
+  }
+
+ protected:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    char *ret = detail::align_ptr(curr_, alignment);
+    if (ret + bytes > limit_)
+      throw std::bad_alloc();
+    curr_ = ret + bytes;
+    return ret;
+  }
+
+  // don't deallocate at all
+  void do_deallocate(void *data, size_t bytes, size_t alignment) override {
+  }
+
+  char *curr_ = nullptr, *limit_ = nullptr;
+};
+
+/**
+ * @brief Monotonic resource which allocates memory from an upstream resource, storing
+ *        the metadata in the same memory blocks as the allocated buffers.
+ *
+ * Monotonic resources don't require manual deallocation of memory.
+ * The lifetime of a monotonic resource is limited and all memory will be deallocated in bulk
+ * when the resource is destroyed.
+ */
+class monotonic_host_resource : public memory_resource {
+ public:
+  explicit monotonic_host_resource(memory_resource *upstream,
+                                   size_t next_block_size = 1024)
+  : upstream_(upstream), next_block_size_(next_block_size) {}
+
+  ~monotonic_host_resource() {
+    free_all();
+  }
+
+  void free_all() {
+    while (curr_block_) {
+      assert(curr_block_->sentinel == sentinel_value && "Memory corruption detected");
+      auto *prev = curr_block_->prev;
+      auto *base = reinterpret_cast<char*>(curr_block_) - curr_block_->usable_size;
+      size_t alloc_size = curr_block_->usable_size + sizeof(block_info);
+      upstream_->deallocate(base, alloc_size, curr_block_->alignment);
+      curr_block_ = prev;
+    }
+  }
+
+ private:
+  void *do_allocate(size_t bytes, size_t alignment) override {
+    char *ret = detail::align_ptr(curr_, alignment);
+    if (ret + bytes > limit_) {
+      while (next_block_size_ < bytes + sizeof(block_info))
+        next_block_size_ += next_block_size_;
+      alignment = std::max(alignment, alignof(block_info));
+      curr_ = static_cast<char*>(upstream_->allocate(next_block_size_, alignment));
+      limit_ = curr_ + next_block_size_ - sizeof(block_info);
+      block_info *prev = curr_block_;
+      if (prev) {
+        assert(prev->sentinel == sentinel_value && "Memory corruption detected");
+      }
+      curr_block_ = reinterpret_cast<block_info*>(limit_);
```

Comment:
So we create blocks only for arbitrary sizes, not for every allocation, right?

Reply:
We create a new block when the previous block can't fit the upcoming allocation.

Reply:
ok